redis from mruby

i110 opened the pull request:
This PR makes it able to access redis from mruby hanlder, like the following:

```
        mruby.handler: |
          redis = H2O::Redis.new(:host => '127.0.0.1', :port => 6379)
          Proc.new do |env|
            set = redis.set('key', 'value')
            get = redis.get('key')
            # redis.disconnect
            [ 200, {}, [set.join, get.join] ]
          end
```

TODOs:
- [x] support more redis features (other commands, authentication, etc..)
- [x] consider preparing specific exception class for IO & procotol errors
- [x] embed mruby source code in `handler/mruby/redis.c` ? (I feel it'll reduce maintainability, so let's discuss)
- [x] timeout
- [x] subscribe

yannick commented:
if you forget to call `.join` on the redis answer it segfaults (at least under macOS)

i110 commented:
@yannick sorry for my late reply, and thank you for noticing me that problem. This is still WIP, and after [This PR](https://github.com/h2o/h2o/pull/1173) gets merged, I'll fix it and address FIXMEs.

yannick commented:
is there a reason `call_method` is restricted to a single argument?
instead of having fixed methods it may be much easier to use symbol or string as arguments and forward all args directly to redis. in that way new commands are automatically supported.

yannick commented:
```
@redis.connect
  	paste = @redis.call(:hget, "myhash", "field1").join
``` 

works fine. however there is a bug when multiple values are sent back from redis: 

```
       @redis.connect
  	paste = @redis.call(:hgetall, "myhash").join
```

crashes
```
Assertion failed: (mrb_array_p(*parent)), function try_parentize, file /lib/handler/mruby/redis.c, line 257.
```

i110 commented:
@yannick thank you for notifying me that. I fixed that bug in b100dd5

> also, is there any reason the connect block is not inside .call?

no there were no reasons (if I recall correctly), so I moved `ensure_connected` into `call` method.

yannick commented:
@i110 could you merge master into here, there are a few conflicts i'm unsure about

yannick commented:
@i110 following handler segfaults:

```
redis = H2O::Redis.new(:host =>"localhost", :port => 6379)

(0...100).each do |nr|
  redis.set("test#{nr}", "X"*300)
end

Proc.new do |env|

puts ENV
[200, {}, [env]]
end

```

yannick commented:
@i110 
- how about supporting the selection of the redis DB?  
- maybe i merged wrongly but when i do a few redis operation in a handler beforehand returning Proc.new i get a segfault

i110 commented:
@yannick Thank you for the suggestions,

> how about supporting the selection of the redis DB?

You can simply select the DB by issueing SELECT command (i.e. `redis.select(your_db_name)` ). document: https://redis.io/commands/select

> maybe i merged wrongly but when i do a few redis operation in a handler beforehand returning Proc.new i get a segfault

At https://github.com/h2o/h2o/pull/1152/commits/019966381e861c1874fabd412c6e9f6fed019460 I fixed an issue which can be related to your segfault. please try again when you get a minute.

And finally, I removed WIP from this PR. Remaining features I want to add in the future are:
* subscribe
  * waiting https://github.com/h2o/h2o/pull/1336 gets merged
* timeout
  * still trying to implement, but it would take a little more time

@kazuho could you review when you have the time?

yannick commented:
great stuff and a really useful feature. 
after ~1M real user sessions, each doing simple redis queries i haven't found any more bugs and it seems to be stable.

i110 commented:
@yannick Thank you for reporting that issue! I fixed it at https://github.com/h2o/h2o/pull/1152/commits/fdc79ee2df66cf839f06c1653b26ba94e43d04b6

tagomoris commented:
Are there any progress? This feature looks so useful for me too :D

i110 commented:
@tagomoris Thanks to you, great progress was made last night.
@kazuho Could you review this when you get a chance?

yannick commented:
there might still be a leak:
valgrind reports:
first a empty mruby handler:
https://gist.github.com/yannick/6f957847bb7fa6d88bca12fd1b792d7d
then a handler that calls `redis.evalsha( "hash").join)`:
https://gist.github.com/yannick/e5436edc5aefca1f18510cded573d422

i110 commented:
Thank you @yannick, I fixed that problem at https://github.com/h2o/h2o/pull/1152/commits/70132d8c7b93dcf3041461b8479c896e0285c568 !

kazuho commented on include/h2o/mruby_.h:
```diff
@@ -117,12 +119,14 @@ typedef struct st_h2o_mruby_generator_t {
     } refs;
 } h2o_mruby_generator_t;
 
+#define H2O_MRUBY_CALLBACK_ID_NOOP 0
```
I am not sure if `0` as a callback ID is allowed. See https://github.com/h2o/h2o/blob/5b71e0a/lib/handler/mruby.c#L773.

i110 replied:
In current implementation, `0` is actually not meaningful and I believe there're no users who actively uses it because it only raises runtime error which means programming error. OTOH we need a way to exit h2o_mruby_run_fiber

kazuho commented on lib/common/redis.c:
```diff
@@ -90,73 +151,267 @@ void h2o_redis_connect(h2o_redis_conn_t *conn, const char *host, uint16_t port)
+        return;
+    }
+
+    if (command->type == H2O_REDIS_COMMAND_TYPE_MONITOR) {
```
Do we need to have code for "monitor" assuming that we always return a error?

kazuho commented on lib/common/redis.c:
```diff
@@ -90,73 +151,267 @@ void h2o_redis_connect(h2o_redis_conn_t *conn, const char *host, uint16_t port)
+        return;
+
+    int err = H2O_REDIS_ERROR_NONE;
+    char *errstr = NULL;
```
Do we need `err` and `errstr`?

I would appreciate it if you could merge them instead of having two. One way would be to only use `const char *` and have specific pointer values defined in the header file. Please consult http1client for details.

i110 replied:
At https://github.com/h2o/h2o/commit/3ead0dac3b178453c9b592aa3b86fffb3f42a16b I remoted `int err`, but still I'm feeling that it may be necessary to report better error message to the client, regarding the fact that error message may came from `strerror`

kazuho commented on lib/common/redis.c:
```diff
@@ -38,21 +59,60 @@ static void on_redis_connect(const redisAsyncContext *redis, int status)
+    h2o_timeout_unlink(entry);
+
+    conn->_did_connect_timeout = 1;
+    invoke_deferred(conn, &conn->_defer_timeout_entry, on_connect_timeout_deferred);
```
I would apppreciate if you could rather not make deferred calls within timeout callbacks. I believe that there is no reason to do so. Doing more things at once is generally better for performance (better locality) and debugging (gives us better trace).

i110 replied:
> I believe that there is no reason to do so.

At that time there're some reasons to do workaround with hiredis async API, but at  https://github.com/h2o/h2o/pull/1152/commits/3ead0dac3b178453c9b592aa3b86fffb3f42a16b I refactored to eliminate such tricks

kazuho commented on lib/common/redis.c:
```diff
@@ -38,21 +59,60 @@ static void on_redis_connect(const redisAsyncContext *redis, int status)
     }
 }
 
+static void on_disconnect(const redisAsyncContext *redis, int status)
```
Does the introduction of the function mean that we now have a "disconnecting" (or "closing") state?

If yes, is there a need to add the state to `h2o_redis_connection_state_t`? If no, why do we need this function?

I would also appreciate it if you could consider unifying the term being used. At the moment I see "close" and "disconnect".

i110 replied:
> Does the introduction of the function mean that we now have a "disconnecting" (or "closing") state?

In our design, disconnecting hiredis object cannot be touched because once disconnecting procedure is fired our redis client immediately detach hiredis object and try to create new connection if another request comes. Regarding that, I don't think we have to have `disconnecting` state.

kazuho commented on lib/common/redis.c:
```diff
@@ -38,21 +59,60 @@ static void on_redis_connect(const redisAsyncContext *redis, int status)
+{
+    h2o_redis_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_redis_conn_t, _defer_timeout_entry, entry);
+
+    assert((conn->_redis->c.flags & REDIS_CONNECTED) == 0);
```
Am I correct in assuming that there is a value that represents the connection state in the `redisAsyncContext`? If so, can we use the value instead of having `h2o_redis_connection_state_t`?

Having two states that represents the same thing is a source of confusion.

i110 replied:
> there is a value that represents the connection state in the redisAsyncContext?

Yes, but it's a bit redundant and seems to be not intended to used directly from outside of hiredis. And also considering that we have a bit different state model described in  https://github.com/h2o/h2o/pull/1152#discussion_r164332530, I think we should have our own state representation.

yannick commented:
it would be great if there would be a way to re-initialize the redis connector if it got killed  
or at least we need some example on how this should be handled.

afair: tearing up the redis connection outside of the request handler proc leaves the handler in a non working state if e.g. redis is restarted.

i110 commented:
@yannick 
You can check whether the connection is still alive or not using `redis.disconnected?` Is this what you need?
OTOH `connect` method internally checks it, and all redis-command method (like `get`) calls `connect` method before emitting commands. So I think the users don't need to use `disconnect?` method in almost cases.

i110 commented on include/h2o/redis.h:
```diff
@@ -38,19 +43,45 @@ typedef struct st_h2o_redis_conn_t {
-    h2o_timeout_entry_t _timeout_entry;
+    h2o_timeout_entry_t _defer_timeout_entry;
+    h2o_timeout_t _connect_timeout;
+    h2o_timeout_entry_t _connect_timeout_entry;
```
Basically the lifecycle of redis object is managed by hiredis internal. For example, if `on_connect` gets called with `REDIS_ERR`, immediately after it hiredis will call `__redisAsyncDisconnect` which will finally dispose redis object.
https://github.com/h2o/h2o/blob/master/deps/hiredis/async.c#L503-L504

i110 commented on lib/common/redis.c:
```diff
@@ -19,40 +19,138 @@
+    if (h2o_timeout_is_linked(&conn->_connect_timeout_entry))
+        h2o_timeout_unlink(&conn->_connect_timeout_entry);
+    if (h2o_linklist_is_linked(&conn->_connect_timeout._link))
+        h2o_timeout_dispose(conn->loop, &conn->_connect_timeout);
```
`conn->_connect_timeout` may be uninitialized if the user set `conn->connect_timeout` to zero. so we have to check it by `h2o_linklist_is_linked(&conn->_connect_timeout._link)`

i110 commented on lib/core/util.c:
```diff
@@ -224,7 +224,7 @@ static h2o_iovec_t build_redis_value(h2o_iovec_t session_data)
 #undef BASE64_LENGTH
 
-static void redis_resumption_on_get(redisReply *reply, void *_accept_data)
+static void redis_resumption_on_get(redisReply *reply, void *_accept_data, const char *errstr)
```
If I recall correctly, `on_redis_resumption_get_failed` is introduced with zero timeout because  `h2o_socket_ssl_resume_server_handshake` or something like that is not reentrant.

i110 commented on lib/handler/mruby/redis.c:
```diff
@@ -0,0 +1,313 @@
+static void on_gc_dispose_command(mrb_state *mrb, void *_ctx)
+{
+    struct st_h2o_mruby_redis_command_context_t *ctx = _ctx;
+    if (ctx == NULL) return;
```
fixed at https://github.com/h2o/h2o/pull/1152/commits/c3d0f8587163607964f85a1401aad3217d3df727

i110 commented on lib/handler/mruby/redis.c:
```diff
@@ -0,0 +1,313 @@
+static void on_gc_dispose_redis(mrb_state *mrb, void *_conn)
+{
+    struct st_h2o_mruby_redis_conn_t *conn = _conn;
+    if (conn == NULL) return;
```
fixed at https://github.com/h2o/h2o/pull/1152/commits/c3d0f8587163607964f85a1401aad3217d3df727

yannick commented:
@i110 ah right, i forgot that it autoconnects, my bad sorry.

kazuho commented:
Thank you for your hard work and your patience!

proyb6 commented:
I have downloaded ```master``` and test to find ```mruby raised: /etc/h2o.conf:112: uninitialized constant H2O::Redis (NameError)``` based on the code example in the first comment at the top of this issue.
